### ISSUE TITLE
Question of the equation in Flash Attention 2 Paper

### ISSUE THREAD

**Issue by jeffrey-sunh1** [Dao-AILab/flash-attention#1349]
Hi Tri, 
@tridao 

I have a question about the equation in chapter 3.1.1 Forward pass of Flash attention 2 paper.

As you mentioned "Only at the every end of the loop do we scale the final  $$\tilde{O}^{(last)}$$ by $$diag(l^{(last)} )^{−1}$$ to get the right output."

Suppose 2 is the last one, I can use $$diag(\ell^{(2)} )^{−1}$$ to scale $$\tilde{O}^{(2)}$$, then we get the following formula:

$$diag(\ell^{(2)} )^{−1}\tilde{O}^{(2)}$$
$$=diag(\ell^{(2)} )^{−1}[diag(\ell^{(1)})^{-1}O^{(1)}+e^{S^{(2)}-m^{(2)}}V^{(2)}]$$
$$=diag(\ell^{(2)} )^{−1}diag(\ell^{(1)})^{-1}O^{(1)}+diag(\ell^{(2)} )^{−1}e^{S^{(2)}-m^{(2)}}V^{(2)}\quad \quad equation-1 $$

The expected result should be 

$$O^{(2)}=diag(\ell^{(1)}\/\ell^{(2)})^{-1}O^{(1)}+diag(\ell^{(2)})^{-1}e^{S^{(2)}-m^{(2)}}V^{(2)}\quad \quad equation-2 $$

The right-hand side of the plus sign in equation 1 is equal to that in equation 2, but the left-hand sides of the plus sign do not match. Could you provide some clarification or guidance on this matter?

**Comment by SimpleTheoryOfTypes** [comment]
The formula for $O_2 should be:

$O_2 = \text{diag}\bigg(\frac{l_2}{l_1}\bigg)^{-1} e^{m_1-m_2} O_1 + \text{diag}(l_2)^{-1} e^{S_2-m_2}  \times V_2 = \text{diag}(l_2)^{-1} e^{S_1-m} V_1 + \text{diag}(l_2)^{-1} e^{S_2-m} V_2 = O$

**Comment by jeffrey-sunh1** [comment]
@SimpleTheoryOfTypes So these are typos in the paper?
![image](https://github.com/user-attachments/assets/71e6a8a1-f277-4d91-bb0a-5e8988f3d83c)

**Comment by SimpleTheoryOfTypes** [comment]
Yes, I believe so. It would be best for the original authors to confirm. @tridao

**Comment by reoLantern** [comment]
> @SimpleTheoryOfTypes So these are typos in the paper? ![image](https://private-user-images.githubusercontent.com/100344695/388772267-71e6a8a1-f277-4d91-bb0a-5e8988f3d83c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMxMjExMjQsIm5iZiI6MTczMzEyMDgyNCwicGF0aCI6Ii8xMDAzNDQ2OTUvMzg4NzcyMjY3LTcxZTZhOGExLWYyNzctNGQ5MS1iYjBhLTVlODk4OGYzZDgzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMlQwNjI3MDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ODg0MmMzY2M4Y2NhMDIxODY4YmExYzgzMzcyNjdhNTNmZDBhNDMyOTg4ODViOTE1Y2FmNDkyN2JiYjQ5MDM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.AYsCgpyqAYx0fWBaN4WEZymGbg8VAZ9EjzpkTom7FLI)

I noticed the same issue while reading the paper. Additionally, since $O^{(i)}$ omits division by $l^{(i)}$ in each iteration, shouldn't the final result be divided by $\prod l^{(i)}$ across all iterations?

**Comment by paluchasz** [comment]
> @SimpleTheoryOfTypes So these are typos in the paper? ![image](https://private-user-images.githubusercontent.com/100344695/388772267-71e6a8a1-f277-4d91-bb0a-5e8988f3d83c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYwNzkyMTAsIm5iZiI6MTczNjA3ODkxMCwicGF0aCI6Ii8xMDAzNDQ2OTUvMzg4NzcyMjY3LTcxZTZhOGExLWYyNzctNGQ5MS1iYjBhLTVlODk4OGYzZDgzYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEwNVQxMjA4MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNDQwODE3YmViZTI4NThmMjU2ZDdkNWU4M2RlNDFlYmJjZTdkZTE0ODZiMzQ2NGZjZDIwMGFhZTA1M2M2MjE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.qJFe3Ezozv6rz4LLSgI5fv61UFCKUraMN9ay49lF3QU)

I also thought these are typos in the paper
